Added .get, renamed .getAll as .getRaw, added new .getAll method

bevacqua opened the pull request:
As pointed out by @bargs in #7301, the `.get` method was missing from the server-side settings config module. I took the liberty of also changing `.getAll` into `.getRaw` since it returns the entire manifest objects for the settings _(includes descriptions, user-provided, and default values)_, and added a new `.getAll` implementation where an object of configuration key/value pairs is provided.

> Note: `.getAll` wasn't being used anywhere anyways.

Bargs commented:
What would you say to adding a cache for these settings? One thing that was unexpected/frustrating when I first tried to use this API was that the `get` methods returned promises. If the entire app is going through this API to make changes, it shouldn't be a problem knowing when to invalidate the cache for a given key.

bevacqua commented:
Sounds good but also would need a hook to learn about changes made directly against the index?

Bargs commented on src/ui/settings/index.js:
```diff
@@ -13,7 +15,23 @@ export default function setupSettings(kbnServer, server, config) {
+  }
+
   function getAll() {
+    return getRaw()
```
We should probably have some tests for these higher level methods.

Bargs commented:
@bevacqua do we still have code that directly accesses the .kibana index? Or are you thinking of user's modifying it themselves outside of Kibana?

bevacqua commented:
I meant users. I was told to trust no one.

Bargs commented:
Haha ok. I think there's a case to be made for users modifying Kibana internal data at their own risk, but it's not a huge deal, I'm ok with leaving it as is if we wanna play it safe.

In that case the code looks good, but I would like to see some tests that exercise those getters at least.

Also one other note, since I *am* using `getAll` in the feature/ingest branch, we'll need to keep that in mind depending on which PR gets merged first. It'll probably be yours and you won't have to worry about it, but in the off chance feature/ingest gets in first we'll need to update that usage in this PR.

bevacqua commented:
I'd say block your issue on this one and I'll add tests so we can merge it before you do yours. Can you x-ref them?

Bargs commented:
Yeah left a note on my PR

bevacqua commented:
@Bargs

- Added a metric ton of tests
- Updated your API usage in `master` [(via `fba84993`)](https://github.com/elastic/kibana/pull/7374/commits/fba849937afe1caf7bf2647dcd180a41b368fd2a)
- Added `removeMany` method

bevacqua commented:
jenkins, test it

cjcenizal commented on src/ui/settings/__tests__/index.js:
```diff
@@ -0,0 +1,345 @@
+    it('updates a single value in one operation', function () {
+      const { server, uiSettings, configGet } = instantiate();
+      const result = uiSettings.setMany({ one: 'value' });
+      expect(typeof result.then).to.be('function');
```
I think these tests can be simplified by creating a separate "returns a promise" test that makes this kind of assertion instead of repeating it within each test. This will make each test's description more accurate, make the tests more readable since they will be testing "just one thing", and the tests themselves will be deflated a bit.

```javascript
describe('#setMany()', () => {
    it('returns a promise', () => {
      const { uiSettings } = instantiate();
      const result = uiSettings.setMany({});
      expect(typeof result.then).to.be('function');
    });

    it('updates a single value in one operation', () => {
      /* ... */
    });

    it('updates several values in one operation', () => {
      /* ... */
    });
  });
```

bevacqua replied:
Totally. Split out assertions into separate tests.

cjcenizal commented on src/ui/settings/__tests__/index.js:
```diff
@@ -0,0 +1,345 @@
+import init from '..';
+import defaultsProvider from '../defaults';
+
+describe('ui settings', function () {
```
I think all the anonymous functions can be replaced with fat arrow functions in this file, to get more ES6ish.

Bargs commented:
jenkins, test it

bevacqua commented:
@Bargs 🏓 _(Ping)_

Bargs commented on src/ui/settings/__tests__/index.js:
```diff
@@ -0,0 +1,307 @@
+    it('returns a promise', () => {
+      const { uiSettings } = instantiate();
+      const result = uiSettings.setMany({ a: 'b' });
+      expect(typeof result.then).to.be('function');
```
Do we know what promise implementation is returned? If so, expect.js's constructor assertion could be used instead of duck typing:

`expect(result).to.be.a(Promise);`

Bargs commented on src/ui/settings/__tests__/index.js:
```diff
@@ -0,0 +1,307 @@
+      const result = await uiSettings.getRaw();
+      const merged = defaultsProvider();
+      merged.foo = { userValue: 'bar' };
+      expect(result).to.eql(merged);
```
I'm super confused. `getRaw` changes the results of `defaultsProvider`?

bevacqua replied:
Nope. `getRaw` pulls a fresh copy of the defaults from a factory.

Bargs commented:
One thing that's not obvious to me from the tests or the documentation is the difference between `getRaw` and `getUserProvided` or why I should prefer one over the other. It seems like `getUserProvided` doesn't merge the config in ES with the defaults, but if that were the case I don't see why the results are returned with nested objects with `userValue` keys. If `getUserProvided` only returns user values, couldn't it just return a flat object?

Bargs commented:
@bevacqua just a few minor questions/comments

bevacqua commented:
@bargs `.getUserProvided()` is leveraged by the REST API to respond with what's changed. The defaults aren't resent to the client as they aren't going to change unless the app gets restarted. That function uses the ES model. The `.getDefaults()` method uses the same model. `.getRaw()` also uses the same model, merging both user configuration in ES and defaults. Then there's `.getAll()`, which takes the output from `.getRaw()` and turns it into key value pairs as you mentioned.

bevacqua commented:
I don't see any test where I could get away not using `.eql`, as strict equality means my expectation would not match, there's new objects pretty much everywhere.

bevacqua commented:
jenkins, test it

Bargs commented:
> I don't see any test where I could get away not using .eql, as strict equality means my expectation would not match, there's new objects pretty much everywhere.

Lodash's _.isEqual only uses strict comparison on object/array properties, not the objects/arrays themselves. So if we had three variables:

`var foo = { something: { userValue: 0 } };`
`var bar = { something: { userValue: '0' } };` // notice this one is a string
`var baz = { something: { userValue: 0 } };`

We would get the following results:

`_.isEqual(foo, bar); // false`
`_.isEqual(foo, baz); // true`

`expect(foo).to.eql(bar); // true`
`expect(foo).to.eql(baz); // true`

Bargs commented:
> .getUserProvided() is leveraged by the REST API to respond with what's changed. The defaults aren't resent to the client as they aren't going to change unless the app gets restarted. That function uses the ES model. The .getDefaults() method uses the same model. .getRaw() also uses the same model, merging both user configuration in ES and defaults. Then there's .getAll(), which takes the output from .getRaw() and turns it into key value pairs as you mentioned.

>The non-flat methods aren't flat because the metadata is used on the client-side, and to maintain the format for easy merging

Could you add some comments in the code to this effect? Otherwise I think someone might misuse these methods on accident.

bevacqua commented:
Oh, wow, I didn't expect `.eql` to be that loose.

Bargs commented:
Yeah, it sucks... that's why I'd still like to [switch to Chai](https://github.com/elastic/kibana/issues/5894).

bevacqua commented:
@Bargs Added comments, used `.isEqual`

bevacqua commented:
jenkins, test it

bevacqua commented:
jenkins, test it

Bargs commented:
jenkins, test it

Bargs commented:
Pending tests passing, LGTM

bevacqua commented:
jenkins, test it

bevacqua commented:
jenkins, test this

cjcenizal commented:
LGTM
